src/usexp/dune_lexer.mll

```diff
@@ -74,57 +8,98 @@ let blank     = [' ' '\t' '\012']
 let digit     = ['0'-'9']
 let hexdigit  = ['0'-'9' 'a'-'f' 'A'-'F']
 
-let atom_char_jbuild =
-  [^ ';' '(' ')' '"' ' ' '\t' '\r' '\n' '\012']
 let atom_char_dune =
   [^ '%' ';' '(' ')' '"' '\000'-'\032' '\127'-'\255']
 
-(* rule for jbuild files *)
-rule jbuild_token = parse
+(* rule for dune files *)
+rule token = parse
   | newline
-    { Lexing.new_line lexbuf; jbuild_token lexbuf }
+    { Lexing.new_line lexbuf; token lexbuf }
   | blank+ | comment
-    { jbuild_token lexbuf }
+    { token lexbuf }
   | '('
     { Token.Lparen }
   | ')'
     { Rparen }
   | '"'
     { Buffer.clear escaped_buf;
       let start = Lexing.lexeme_start_p lexbuf in
-      let s = quoted_string Old_syntax lexbuf in
+      let s = dune_quoted_string lexbuf in
       lexbuf.lex_start_p <- start;
       Quoted_string s
     }
-  | "#|"
-    { jbuild_block_comment lexbuf;
-      jbuild_token lexbuf
-    }
-  | "#;"
-    { Sexp_comment }
+  | atom_char_dune+ as s
+    { Token.Atom (Atom.of_string s) }
+  | _ as c { error lexbuf (Printf.sprintf "Invalid atom character '%c'" c) }
   | eof
     { Eof }
+
+and dune_quoted_string = parse
+  | "\\|"
+    { block_string_start With_escape_sequences lexbuf }
+  | "\\>"
+    { block_string_start Raw lexbuf }
   | ""
-    { jbuild_atom "" (Lexing.lexeme_start_p lexbuf) lexbuf }
+    { quoted_string New_syntax lexbuf }
```

Comment:
The `New_syntax` mode should no longer be necessary since this file only parses the new syntax